Replace embedded-time with fugit

andresv opened the pull request:
This PR replaces `embedded-time` with more lightweight const generics based time library called [fugit](https://lib.rs/crates/fugit) which is for example going to be used in RTIC.

`embedded-time` had some dependencies that did not work well on ESP32, this PR fixes this particular problem.

andresv committed:
Cargo.toml: replace embedded-time with fugit

And use ublox-sockets that support fugit time crate.

andresv committed:
lib: introduce new Clock trait and Instant alias

andresv committed:
replace embedded-time with fugit

`embedded-time` has some serious issues and moves too slowly and tries to do too much in single lib. Fugit is a time library that was built because some people were not happy to use embedded-time on embedded systems.
Fugit only deals with time conversions and those conversions are done using const generics. So most of the stuff is  happening during compilation.
This is very light weight compared to embedded-time.

andresv commented on ublox-cellular/src/lib.rs:
```diff
@@ -24,6 +24,26 @@ pub use services::data::DataService;
 // Re-export atat version in use
 pub use atat;
 
+pub type Instant<const FREQ_HZ: u32> = fugit::TimerInstantU32<FREQ_HZ>;
+
+#[derive(Debug, PartialEq)]
+pub enum ClockError {
+    Infallible,
+}
+
+/// `Clock` provides all timing capabilities that are needed for the library.
+///
+/// Notice that `Clock` trait uses [fugit](https://lib.rs/crates/fugit) crate for `Duration` and `Instant`.
+pub trait Clock<const FREQ_HZ: u32> {
+    fn now(&mut self) -> fugit::TimerInstantU32<FREQ_HZ>;
+
+    fn start<T>(&mut self, count: T) -> Result<(), ClockError>
+    where
+        T: Into<fugit::MillisDurationU32>;
+
+    fn wait(&mut self) -> Result<(), ClockError>;
+}
+
```
`fugit` only provides `Duration` and `Instant` and that is it. Therefore here new `Clock` trait is introduced to work with those types.

This is rough sketch what it should do, maybe it can be simplified. Because essentially we just need `now` and something to make  a `delay`.

andresv replied:
Note - It might be better to use those types more like this: https://github.com/rtic-rs/systick-monotonic/blob/new-trait-and-time-lib/src/lib.rs#L36-L48

andresv commented:
Ahaa, I think I know why I did not see these errors. I added `std` to:

```
atat = { version = "0.13.1", features = ["std", "derive", "defmt", "bytes"] }
```

andresv commented:
I just realized that `atat` uses `embedded-hal` `CountDown` trait (https://github.com/BlackbirdHQ/atat/blob/master/atat/src/client.rs#L58), but at the moment here we are using our own `Clock` trait.

It would be actually better if we use `CountDown` also here, but we also need `now()` method to get current time and I guess there is also problem with time conversion. Typically `CountDown` is implemented for `u32`, but here we would like to use it more like for `fugit::MillisDurationU32`.

Lemme think about it some more.

MathiasKoch commented:
Yeah . The timer stuff across the stack is not ideal, so i am very much open to solutions that would make it easier to use..

andresv committed:
rename FREQ_HZ to TIMER_HZ

the same is used in rtic and it is more clear what it is

andresv committed:
lib: export Clock from atat

andresv committed:
use Clock trait from atat lib

andresv commented:
`Clock` trait now comes from `atat`: https://github.com/andresv/atat/blob/use-fugit-for-timing/atat/src/lib.rs#L278-L290

MathiasKoch commented:
This looks amazing! Awesome job so far! Hope i can find the time to review both atat & this tomorrow 🤞👌

andresv committed:
examples: update common_lib to latest

andresv committed:
examples: linux wip

MathiasKoch committed:
Make sure the embedded-nal implementations only makes use of error-ty… (#53)

* Make sure the embedded-nal implementations only makes use of error-types from ublox-sockets

* Add embedded-hal-driver keyword to crates.io
# Conflicts:
#	ublox-cellular/Cargo.toml
#	ublox-cellular/src/services/data/tcp_stack.rs
#	ublox-cellular/src/services/data/udp_stack.rs

andresv commented:
`Clock` in `atat` was changed to use `WouldBlock` for wait. This is not yet updated here.

andresv commented:
Hmm, I guess cherry-picking your latest commit was not right thing to do (this conflict message here).

andresv committed:
test_helpers: update to WouldBlock based Clock

andresv committed:
error: add helper to convert Clock error to Error::Generic

From trait cannot be used here because Clock error is associated type

andresv committed:
client: use nb::block to wait for timer

andresv committed:
import atat from blackbirdhq master

andresv commented:
`Clock` is now fixed.

MathiasKoch committed:
Merge branch 'master' into replace-embedded-time-with-fugit

andresv commented:
It makes most sense if Format is implemented for fugit types: https://github.com/korken89/fugit/issues/6

andresv committed:
use fugit defmt feature to get Format impl

andresv committed:
linux: set_socket_storage

andresv committed:
add IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr to prelude

andresv committed:
linux example: add socket read/write

andresv commented:
Linux example builds now: https://github.com/andresv/ublox-cellular-rs/blob/replace-embedded-time-with-fugit/examples/linux/src/main.rs

andresv committed:
reexport fugit

fugit is needed for implementing Clock trait therefore it makes sense to reexport it here

andresv committed:
add some general documentation about the crate

andresv committed:
make first examples buildable also on macos

on macos dummy gpio support is used

andresv committed:
rename linux example to sockets, add args support

MathiasKoch commented:
Can you update this PR to make use of atat 0.14.0 & ublox-sockets 0.1.0 releases as well?

andresv commented:
Yes. Adding those and starting to use `fmt.rs` file for logging.

andresv commented:
I wonder how to enable `log` or `defmt` feature for `atat` and `ublox-sockets` based on if `log` or `defmt` feature is selected for this crate.
For example here `log` is at the moment hardcoded.

```toml
[dependencies]
atat = { version = "0.14", features = ["derive", "log", "bytes"] }
ublox-sockets = { version = "0.1", features = ["log"] }

log = { version = "^0.4", default-features = false, optional = true }
defmt = { version = "^0.2", optional = true }

[features]
default = ['log", "toby-r2", "socket-udp", "socket-tcp"]

lara-r2 = []
leon-g1 = []
lisa-u2 = []
mpci-l2 = []
sara-g3 = []
sara-g4 = []
sara-u1 = []
sara-u2 = ["upsd-context-activation"]
toby-l2 = []
toby-r2 = []
topy-l4 = []

upsd-context-activation = []

socket-tcp = ["ublox-sockets/socket-tcp"]
socket-udp = ["ublox-sockets/socket-udp"]

defmt-default = ["defmt"]
defmt-trace = ["defmt"]
defmt-debug = ["defmt"]
defmt-info = ["defmt"]
defmt-warn = ["defmt"]
defmt-error = ["defmt"]
```

andresv committed:
power: add nb::block! for blocking timer.waits

without those it returns immediately without blocking in place

andresv committed:
use atat and ublox-sockets from crates.io

andresv committed:
use fmt.rs to support log or defmt

andresv commented:
FYI: this is not dropped, I am continuing with it quite soon.

MathiasKoch commented:
No problem :+1: 
I have not had a chance to replace our own embedded-time stuff with fugit yet, so i am not in a rush.. Hopefully i'll find the time in the beginning of the new year.

Let med know if you need anything from me :) 

Merry Christmas :christmas_tree:

andresv commented:
Finally back in hacking with it.

andresv committed:
examples: add better logging

MathiasKoch commented:
https://github.com/BlackbirdHQ/atat/pull/113 merged

MathiasKoch reviewed (changes requested):
Would you mind updating this to `embedded-hal: "^1.0.0-alpha.6"` and `defmt: "^0.3", now that https://github.com/BlackbirdHQ/atat/pull/113 is merged?

andresv committed:
update to embedded-hal 1.0.0-alpha.6

andresv committed:
add .DS_Store to .gitignore

macOS creates those nasty files

andresv commented:
Updated, but waiting for new `atat` release: https://github.com/BlackbirdHQ/atat/issues/116.

MathiasKoch commented:
Yeah, i also have a few PRs blocked on that, so it will be out during today.

Just want to check if https://github.com/BlackbirdHQ/atat/pull/115 makes any significant code-size changes first

MathiasKoch commented:
`atat v0.15.0` & `ublox-sockets v0.2.0` released :tada:

andresv committed:
update atat and ublox-sockets

MathiasKoch committed:
Merge branch 'master' into replace-embedded-time-with-fugit

MathiasKoch commented on ublox-cellular/Cargo.toml:
```diff
@@ -15,22 +15,23 @@ name = "ublox_cellular"
 serde = { version = "^1", default-features = false, features = ["derive"] }
-ublox-sockets = { git = "https://github.com/BlackbirdHQ/ublox-sockets", rev = "61413b7" }
-# ublox-sockets = { path = "../../ublox-sockets" }
+ublox-sockets = { version = "0.2" }
```
```suggestion
ublox-sockets = { version = "0.2", features = ["defmt"] }
```

MathiasKoch committed:
Update ublox-cellular/Cargo.toml

MathiasKoch commented:
Merged here: https://github.com/BlackbirdHQ/ublox-cellular-rs/pull/57
